nexus/db-queries/src/db/datastore/disk.rs

```diff
@@ -771,117 +787,287 @@ impl DataStore {
+                                | api::external::InstanceState::Stopped => {
                                     // The disk is ready to be attached, and the
                                     // instance is ready to be attached. Perhaps
                                     // we are at attachment capacity?
                                     if attached_count == i64::from(max_disks) {
-                                        return Err(Error::invalid_request(&format!(
-                                            "cannot attach more than {} disks to instance",
-                                            max_disks
-                                        )));
+                                        return Err(Error::invalid_request(
+                                            &format!(
+                                                "cannot attach more than \
+                                                {max_disks} disks to instance",
+                                            ),
+                                        ));
+                                    }
+
+                                    // Was this attach of a local storage disk
+                                    // blocked due to an existing sled resource
+                                    // record?
+                                    if local_storage_disk {
+                                        use sled_resource_vmm::dsl;
+
+                                        let record = dsl::sled_resource_vmm
+                                            .filter(
+                                                dsl::instance_id
+                                                    .eq(authz_instance.id()),
+                                            )
+                                            .select(dsl::instance_id)
+                                            .execute_async(&*conn)
+                                            .await
+                                            .optional()
+                                            .map_err(|e| {
+                                                public_error_from_diesel(
+                                                    e,
+                                                    ErrorHandler::Server,
+                                                )
+                                            })?;
+
+                                        if record.is_some() {
+                                            let s = "cannot attach local \
+                                                storage disk: sled reservation \
+                                                has already occurred for \
+                                                instance";
+
+                                            return Err(
+                                                Error::invalid_request(
+                                                    &s.to_string(),
```

Comment:
`s` is already an `&str` here, writing `&s.to_string()` is...allocating a `String` just to pass a reference to that string to `invalid_request`, which will allocate a _second_ `String` for it and throw the first one away. isn't this just:
```suggestion
                                                    s,
```